synapse/storage/databases/state/bg_updates.py

```diff
@@ -98,153 +98,100 @@ def _get_state_groups_from_groups_txn(
-                for etype, skey in state_filter_condition_combos:
-                    if skey is None:
-                        where_clause = "(type = ?)"
-                        overall_select_query_args.extend([etype])
-                    else:
-                        where_clause = "(type = ? AND state_key = ?)"
-                        overall_select_query_args.extend([etype, skey])
-
-                    select_clause_list.append(
-                        f"""
-                        (
-                            SELECT DISTINCT ON (type, state_key)
-                                type, state_key, event_id
-                            FROM state_groups_state
-                            INNER JOIN sgs USING (state_group)
-                            WHERE {where_clause}
-                            ORDER BY type, state_key, state_group DESC
-                        )
-                        """
+        # This is an optimization to create a select clause per-condition. This
+        # makes the query planner a lot smarter on what rows should pull out in the
+        # first place and we end up with something that takes 10x less time to get a
+        # result.
+        use_condition_optimization = (
+            not state_filter.include_others and not state_filter.is_full()
+        )
+        state_filter_condition_combos: List[Tuple[str, Optional[str]]] = []
+        # We don't need to caclculate this list if we're not using the condition
+        # optimization
+        if use_condition_optimization:
+            for etype, state_keys in state_filter.types.items():
+                if state_keys is None:
+                    state_filter_condition_combos.append((etype, None))
+                else:
+                    for state_key in state_keys:
+                        state_filter_condition_combos.append((etype, state_key))
+        # And here is the optimization itself. We don't want to do the optimization
+        # if there are too many individual conditions. 10 is an arbitrary number
+        # with no testing behind it but we do know that we specifically made this
+        # optimization for when we grab the necessary state out for
+        # `filter_events_for_client` which just uses 2 conditions
+        # (`EventTypes.RoomHistoryVisibility` and `EventTypes.Member`).
+        if use_condition_optimization and len(state_filter_condition_combos) < 10:
+            select_clause_list: List[str] = []
+            for etype, skey in state_filter_condition_combos:
+                if skey is None:
+                    where_clause = "(type = ?)"
+                    overall_select_query_args.extend([etype])
+                else:
+                    where_clause = "(type = ? AND state_key = ?)"
+                    overall_select_query_args.extend([etype, skey])
+
+                select_clause_list.append(
+                    f"""
+                    (
+                        SELECT DISTINCT ON (type, state_key)
```

Comment:
https://github.com/matrix-org/synapse/actions/runs/3529532667/jobs/5920661864#step:9:11649

```
sqlite3.OperationalError: near "ON": syntax error
```

Reply:
I think we can just use `LIMIT 1` instead because we are only picking one condition at a time which targets a single event.

Whereas in the non-optimized version, we want to get multiple events out of the same query

Reply:
Related https://stackoverflow.com/a/71924314/796832

Reply:
> I think we can just use `LIMIT 1` instead because we are only picking one condition at a time which targets a single event.

Ended up using this for the optimized query :fast_forward:. Seems simpler than `SELECT DISTINCT ON` anyway for this.

There is another road-block here though with not being able to use `LIMIT 1` though -> https://github.com/matrix-org/synapse/pull/14531#discussion_r1191853344


> But that still leaves the `SELECT DISTINCT ON` in the non-optmized version 🤔

I couldn't figure out a way to make something better and compatible with SQLite so I went with a Postgres/SQLite specific versions of the query and just filter out the duplicates. We do still get the benefit of using the same recursive query stuff now and same row processing though.

See https://github.com/matrix-org/synapse/pull/14531/commits/8b103f46a7b9959f4dcf7ea64ff6fabcf292a854